extensions/wikia/Insights/models/InsightsQuerypageModel.php

```diff
@@ -37,17 +52,132 @@ public function getData() {
 	 *
 	 * @return array
 	 */
-	public function getContent() {
+	public function getContent( $params ) {
 		$this->queryPageInstance = $this->getDataProvider();
 
+		/**
+		 * 1. Prepare data of articles - title, last revision, link etc.
+		 */
+		$articlesData = $this->fetchArticlesData();
+
+		/**
+		 * 2. Slice a sorting table to retrieve a page
+		 */
+		$this->prepareParams( $params );
+		if ( !isset( $this->sortingArray ) ) {
+			$this->sortingArray = array_keys( $articlesData );
+		}
+		$ids = array_slice( $this->sortingArray, $this->offset, $this->limit, true );
+
+		/**
+		 * 3. Populate $content array with data for each article id
+		 */
 		$content = [];
-		$res = $this->queryPageInstance->doQuery( $this->offset, $this->limit );
-		if ( $res->numRows() > 0 ) {
-			$content = $this->prepareData( $res );
+		foreach ( $ids as $id ) {
+			$content[] = $articlesData[$id];
 		}
 		return $content;
 	}
 
+	private function prepareParams( $params ) {
+		global $wgMemc;
+
+		if ( isset( $params['sort'] ) && isset( $this->sorting[ $params['sort'] ] ) ) {
+			$this->sortingArray = $wgMemc->get( wfSharedMemcKey( $this->getInsightType(), $params['sort'] ) );
+		}
+
+		if ( isset( $params['offset'] ) ) {
+			$this->offset = intval( $params['offset'] );
+		}
+
+		if ( isset( $params['limit'] ) ) {
+			$this->limit = intval( $params['limit'] );
+		}
+	}
+
+	public function fetchArticlesData() {
+		$cacheKey = wfSharedMemcKey( $this->getInsightType(), 'articlesData' );
```

Comment:
Please add wiki id to memcache key. What do you think about having method which returns the key?